core/graph/src/main/java/uk/gov/gchq/gaffer/graph/hook/FunctionAuthoriser.java

```diff
@@ -0,0 +1,209 @@
+ * Functions which have been banned. The Authoriser can be configured with a
+ * list of unauthorised function classes, or a list of patterns to check against.
+ * <p>
+ * It should be noted that using the unauthorisedFunctions list will be more
+ * efficient than using any of the patterns.
+ */
+@JsonPropertyOrder(value = {"unauthorisedFunctions", "unauthorisedFunctionPatterns"})
+@JsonInclude(JsonInclude.Include.NON_DEFAULT)
+public class FunctionAuthoriser implements GraphHook {
+
+    private static final String ERROR_MESSAGE_PREFIX = "Operation chain contained an unauthorised function: ";
+    private static final Logger LOGGER = LoggerFactory.getLogger(FunctionAuthoriser.class);
+    private List<Class<? extends Function>> unauthorisedFunctions = new ArrayList<>();
+    private List<Pattern> authorisedFunctionPatterns = new ArrayList<>();
+    private List<Pattern> unauthorisedFunctionPatterns = new ArrayList<>();
+
+    public FunctionAuthoriser() {
+    }
+
+    public FunctionAuthoriser(final List<Class<? extends Function>> unauthorisedFunctions, final List<Pattern> unauthorisedFunctionPatterns, final List<Pattern> authorisedFunctionPatterns) {
+        this.setUnauthorisedFunctions(unauthorisedFunctions);
+        this.setAuthorisedFunctionPatterns(authorisedFunctionPatterns);
+        this.setUnauthorisedFunctionPatterns(unauthorisedFunctionPatterns);
+    }
+
+    @Override
+    public void preExecute(final OperationChain<?> opChain, final Context context) {
+        String chainString;
+        try {
+            chainString = new String(JSONSerialiser.serialise(opChain));
+        } catch (final SerialisationException e) {
+            // This should never happen in real life as operation chains should
+            // always be json serialisable. However this could happen if using a
+            // mock in testing. To account for this, it will be logged.
+            LOGGER.warn("Failed to serialise operation chain: " + opChain);
+            return;
+        }
+
+        if (unauthorisedFunctions != null) {
+            checkNoBlacklistedFunctionsArePresent(chainString);
+        }
+        if (authorisedFunctionPatterns != null || unauthorisedFunctionPatterns != null) {
+            checkAllFunctionsUsedAppearInPatterns(chainString);
+        }
+    }
+
+    private List<Predicate> convertPatternsToPredicates(final List<Pattern> patterns) {
+        return patterns.stream()
+                .map(Pattern::asPredicate)
+                .collect(Collectors.toList());
+    }
+
+    private void checkAllFunctionsUsedAppearInPatterns(final String chainString) {
+        boolean noWhitelistPatterns = (authorisedFunctionPatterns == null || authorisedFunctionPatterns.isEmpty());
+        boolean noBlacklistPatterns = (unauthorisedFunctionPatterns == null || unauthorisedFunctionPatterns.isEmpty());
+
+        Predicate<String> isAuthorised = createAuthorisationFunction(noBlacklistPatterns, noWhitelistPatterns);
+
```

Comment:
Matching of the "class":" string feels a bit fragile to me. You may want to consider creating a TreeModel from the Json string and traversing the tree of JsonNode objects checking for "class" fields.

Reply:
Rather than going to String then to JsonNode, you might be able to use the valueToTree method on ObjectMapper. Just had a quick look at it and it should work. You would need to expose this in the Gaffer JSONSerialiser class.

Reply:
Would this get around the issue of being reliant on things being serialised as "class" fields?

Reply:
From what I can tell, this would only provide a small performance benefit if you don't make use of the more efficient blacklisting capability (Since the blacklisting capability requires it to be a string anyway).